src/openfermion/ops/_symbolic_operator.py

```diff
@@ -0,0 +1,301 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""SymbolicOperator is a base class for QubitOperator and FermionOperator"""
+import copy
+import itertools
+
+import numpy
+from abc import abstractmethod
+
+
+EQ_TOLERANCE = 1e-12
+
+
+class SymbolicOperatorError(Exception):
+    pass
+
+
+class SymbolicOperator(object):
+    """
+    The base class for QubitOperator and FermionOperator. All methods defined
+    here can be accessed from FermionOperator or QubitOperator objects.
+    This is an abstract class and objects of this type cannot be created,
+    only those of subclasses.
+
+    Subclasses are sums of terms of operators for a particular category of
+    particle. Subclasses support addition and multiplication with objects of
+    the same type.
+
+    Attributes:
+        terms (dict):
+            **key** (tuple of tuples): Each tuple represents a fermion term,
```

Comment:
This description is a bit too narrow since it seems to describe FermionOperators. I think we want something more general which would also encompass QubitOperators. Kevin had a good suggestion to make the abstract class (index, action), etc.